src/resources.rs

```diff
@@ -0,0 +1,351 @@
+    MmioAddress {
+        /// Allocate resource within the range [`min`, `max`] if 'min' is not u64::MAX.
+        min: u64,
+        /// Allocate resource within the range [`min`, `max`] if 'max' is not zero.
+        max: u64,
+        /// Alignment for the allocated address.
+        align: u64,
+        /// Size for the allocated address range.
+        size: u64,
+    },
+    /// Request for a legacy IRQ.
+    LegacyIrq {
+        /// Reserve the preallocated IRQ if it's not u32::MAX.
+        fixed: u32,
+    },
+    /// Request for PCI MSI IRQs.
+    PciMsiIrq {
+        /// Number of Irqs to allocate.
+        size: u32,
+    },
+    /// Request for PCI MSIx IRQs.
+    PciMsixIrq {
+        /// Number of Irqs to allocate.
+        size: u32,
+    },
+    /// Request for generic IRQs.
+    GenericIrq {
+        /// Number of Irqs to allocate.
+        size: u32,
+    },
+    /// Request for KVM mem_slot indexes to map memory into the guest.
+    KvmMemSlot {
+        /// Allocate the specified kvm memory slot index if it's not u32::MAX.
+        fixed: u32,
+        /// Number of slots to allocate.
+        size: u32,
+    },
+}
+
+impl ResourceRequest {
+    /// Create a new Mmio address request object with default configuration.
+    pub fn new_pio(size: u16) -> Self {
+        ResourceRequest::PioAddress {
+            min: u16::MAX,
+            max: 0,
+            align: 0x1,
+            size,
+        }
+    }
+
+    /// Create a new Mmio address request object with default configuration.
+    pub fn new_mmio(size: u64) -> Self {
+        ResourceRequest::MmioAddress {
+            min: u64::MAX,
+            max: 0,
+            align: 0x1000,
+            size,
+        }
+    }
+
+    /// Create a new legacy IRQ request object with default configuration.
+    pub fn new_legacy_irq() -> Self {
+        ResourceRequest::LegacyIrq { fixed: u32::MAX }
+    }
+
+    /// Create a new KVM memory slot request object with default configuration.
+    pub fn new_kvm_mem_slot(size: u32) -> Self {
+        ResourceRequest::KvmMemSlot {
+            fixed: u32::MAX,
+            size,
+        }
+    }
+}
+
+/// Type of Message Singaled Interrupt
+#[derive(Copy, Clone, PartialEq)]
+pub enum MsiIrqType {
+    /// PCI MSI IRQ numbers.
+    PciMsi,
+    /// PCI MSIx IRQ numbers.
+    PciMsix,
+    /// Generic MSI IRQ numbers.
+    GenericMsi,
+}
+
+/// Enumeration for device resources.
+#[allow(missing_docs)]
+#[derive(Clone)]
+pub enum ResourceEntry {
+    /// IO Port address range.
+    PioAddressRange { base: u16, size: u16 },
+    /// Memory Mapped IO address range.
+    MmioAddressRange { base: u64, size: u64 },
+    /// Legacy IRQ number.
+    LegacyIrq(u32),
+    /// Message Signaled Interrupt
+    MsiIrq {
+        ty: MsiIrqType,
+        base: u32,
+        size: u32,
+    },
+    /// Network Interface Card MAC address.
+    MacAddresss(String),
+    /// KVM memslot index.
+    KvmMemSlot(u32),
+}
+
```

Comment:
Using Option<x> to encode fixed resource request is a good idea, but using it to encode allocated resource may seem inconvenient to the resource consumer. We have such experience when implementing tempdir in vmm-sys-util crate.

Reply:
So would something like that:

```Rust
/// Resources flags
pub const MMIO_32BIT: u32 = 1;
pub const MMIO_64BIT: u32 = 2;

/// Type of Message Singaled Interrupt
#[derive(Copy, Clone, PartialEq)]
pub enum MsiIrqType {
    /// PCI MSI IRQ numbers.
    PciMsi,
    /// PCI MSIx IRQ numbers.
    PciMsix,
    /// Generic MSI IRQ numbers.
    GenericMsi,
}

/// A Resource can be used by a device for requesting a specific resource,
/// and by the VMM to pass reserved resources back to a device.
pub enum Resource {
    /// Port IO (PIO) address range.
    PioRange {
        /// In order to request a PIO range that starts at a specific address,
        /// a device must set `base` to such address. Otherwise it should be set
        /// to u16::MAX.
        base: u16,
        /// Size for the allocated address range.
        size: u16,
        /// Alignment for the allocated range base address.
        align: u16,
    }

    /// Memory Mapped IO (MMIO) address range.
    MmioRange {
        /// In order to request an MMIO range that starts at a specific address,
        /// a device must set `base` to such address. Otherwise it should be set
        /// to u64::MAX.
        base: u64,
        /// flags is used by devices to express additional MMIO range resource
        /// requirements, like e.g. the fact that a range must live in the
        /// 32-bit address space.
        flags: u64,
        /// Size for the allocated address range.
        size: u64,
        /// Alignment for the allocated range base address.
        align: u64,
    }

    /// Legacy IRQ number
    LegacyIrq {
        /// In order to request for a specific legacy IRQ number, a device must
        /// set `base` to such IRQ number. Otherwise it should be set to
        /// u32::MAX.
        base: u32,
    },

    /// Message Signaled Interrupt
    MsiIrq {
        /// MSI IRQ type
        msi_type: MsiIrqType,
        /// base must be set to u32::MAX for a device resource request.
        base: u32,
        /// Number of IRQs in the MSI range.
        size: u32,
    },

    /// Network Interface Card MAC address.
    MacAddresss(String),

    KvmMemSlotRange {
        /// In order to request a KVM slot range that starts from a specific
        /// slot index, a device must set `index` to such slot index. Otherwise
        /// it should be set to u32::MAX.
        index: u32,
        /// Size for the allocated slot range.
        size: u32,
    },
}    

```

look better to you?

Reply:
@jiangliu Any feedback on the counter proposal?

Reply:
We could use the same underline data structure to represent requests and assigned resources. On the other handle, it will be much more convenient to provide two sets of methods to create resource request objects and assigned resource objects, such as:
Resource::new_mmio() for assigned mmio resource and resource::new_mmio_request() for mmio request.

Reply:
@jiangliu Assuming we have a `Resource` trait that devices would implement:

```Rust
pub Trait DeviceResources: Send {
     fn get_resources(&self) -> Vec<Resource>;
     fn set_resources(&mut self, Vec<Resource>);
}
```

having one single set of API for allocate and create `Resource`s would mean both the device and the allocator could use the same set of API. Which sounds simpler to me, but I may be missing something?

Reply:
We are not following the typical get_resources/set_resources model. Sample code for MMIO Virtio device as below: 

```
// create the virtio backend driver
let device: Box<dyn VirtioDevice> = Box::new(VirtioBlkDevice::new());

// add resource requests for transport layer        
// Every emulated Virtio MMIO device needs a 4K configuration space.
const MMIO_ADDRESS_DEFAULT: DeviceResourceRequest = DeviceResourceRequest::MmioAddress {
            fixed: std::u64::MAX,
            min: std::u64::MAX,
            max: std::u64::MAX,
            align: 0,
            size: MMIO_DEFAULT_CFG_SIZE,
};
let mut requests = vec![MMIO_ADDRESS_DEFAULT];

// Add resource requests for backend drivers        
device.get_resource_requirements(&mut requests);

// Allocate resources from the resource manager        
let resources = self
            .res_allocator
            .allocate_device_resources(&requests)
            .ok_or(StartMicrovmError::AllocateResource)?;

// Create the MMIO virtio deivce with assigned resources
 let mmio_device = devices::virtio::mmio::MmioDevice::new(
            device,
            resources,
            self.vm_fd.clone(),
            guest_mem.clone(),
            self.irq_manager.clone(),
)
.map_err(|x| Error::CreateDevice(std::io::Error::from_raw_os_error(x.errno())));

// the mmio transport driver will pass the assigned resources to virtio backend drivers when calling
// VirtioDevice::activate(resources).
```

Reply:
This design is model after the Linux resource subsystem, `struct ResourceRequest` corresponds to `struct resource_constraint` and struct Resource corresponds to `struct resource`. The linux resource allocator is flexible enough:)

Reply:
> This design is model after the Linux resource subsystem, `struct ResourceRequest` corresponds to `struct resource_constraint` and struct Resource corresponds to `struct resource`. The linux resource allocator is flexible enough:)

Thanks for the insight of where this is coming from. With the code flow sample, it makes more sense to me now.
What you call `ResourceRequest` is really a resource constraint. Could we please rename it to `ResourceConstraint` then? And rename `ResourceEntry` to simply `Resource`?

Reply:
I have used the term "Constraint" in the PR https://github.com/rust-vmm/vm-allocator/pull/3 for vm-allocator crate.
And I will rename these structs as suggested:)

Reply:
> @jiangliu Assuming we have a `Resource` trait that devices would implement:
> 
> ```rust
> pub Trait DeviceResources: Send {
>      fn get_resources(&self) -> Vec<Resource>;
>      fn set_resources(&mut self, Vec<Resource>);
> }
> ```
> 
> having one single set of API for allocate and create `Resource`s would mean both the device and the allocator could use the same set of APIs. Which sounds simpler to me, but I may be missing something?

@sameo I think It's good to have such trait than implementing all request resources in VMM. According to the PR it seems we need to change the return type a little bit as follows.
```rust
pub Trait DeviceResources: Send {
     // Device object returns the ResourceConstraint to VMM
     fn get_resources_requirement(&self) -> Vec<ResourceConstraint>;
     // Device object sets back the assigned Resource
     fn set_assigned_resources(&mut self, Vec<Resource>);
}
```

Reply:
According to my experience, set_assigned_resources() seems good at design but may not be the best way for implementation.
For example, the assigned resources are not available when creating the device, and will be assigned later. So the device driver must use field "Option<DeviceResources>" to store assigned resources. And you know, Option<> is not very friend.

We have tried the way to "set_assigned_resources()" but give up.

Reply:
> For example, the assigned resources are not available when creating the device, and will be assigned later. So the device driver must use field "Option" to store assigned resources. 

Do you mean that when we create a device object, there is no resources available? How about setting to a MAX value as a hint? In this way, it is used to construct a right Device::ResourceConstraint

Reply:
> For example, the assigned resources are not available when creating the device, and will be assigned later. So the device driver must use field "Option" to store assigned resources. And you know, Option<> is not very friend.

I agree with what you describe here, and it's better to let the device implementation assume that resources are created and available when we create it. But I still believe that there should be a way to asynchronously assign resources to a device, for device reprogramming for example.

Reply:
The PCI bar reprogram is not covered yet. We may solve it later when the vm-allocator becomes more solid.